src/types/network.rs

```diff
@@ -9,255 +9,343 @@ extern crate quickcheck;
+                // the total number of contributions of the account.
+                let total_account_contrib = source_node.data().total_contributions;
+                let user_contribs = edge.data().contributions;
+
+                match (user_contribs, total_account_contrib) {
+                    (None, None) => panic!(
+                        r###"Graph construction incorrect: found a contrib* node
+                        without contributions and an edge without contributions."###
+                    ),
+                    (Some(_), None) => panic!(
+                        r###"Graph construction incorrect: found a contrib* node
+                        with contributions, but an edge without contributions."###
+                    ),
+                    (None, Some(_)) => panic!(
+                        r###"Graph construction incorrect: found a contrib* node
+                        without contributions, but an edge with contributions."###
+                    ),
+                    (Some(uc), Some(total)) => weight = weight * Weight::new(uc, total).into(),
+                }
+            }
+            types::EdgeType::Maintain => {
+                // The weight is divided by the corresponding count of
+                // outgoing edges of the same type on the node.
+                weight = weight / others.into()
+            }
+            types::EdgeType::MaintainStar => {
+                // contrib* and maintain* are multiplied by the number of
+                // contributions of the account to the project, divided by
+                // the total number of contributions of the account.
+                let total_account_contrib = source_node.data().total_contributions;
+                let user_contribs = edge.data().contributions;
+
+                match (user_contribs, total_account_contrib) {
+                    (None, None) => panic!(
+                        r###"Graph construction incorrect: found a contrib* node
+                        without contributions and an edge without contributions."###
+                    ),
+                    (Some(_), None) => panic!(
+                        r###"Graph construction incorrect: found a contrib* node
+                        with contributions, but an edge without contributions."###
+                    ),
+                    (None, Some(_)) => panic!(
+                        r###"Graph construction incorrect: found a contrib* node
+                        without contributions, but an edge with contributions."###
+                    ),
+                    (Some(uc), Some(total)) => weight = weight * Weight::new(uc, total).into(),
+                }
+            }
+            types::EdgeType::Depend => {
+                // The weight is divided by the corresponding count of
+                // outgoing edges of the same type on the node.
+                weight = weight / others.into()
+            }
+        }
+
+        weight
```

Comment:
Depending on how the `dynamic_weight` is used, I am wondering if it is correct. The weights of all outgoing edges should add up to 1, but if a node does not have all types of edges, this wont be the case here. The hyper params would need to be normalized again.

Reply:
Good observation! Initially I was going to normalise everything at the end, but I was taken in by the fact the weights of the spec graph were all summing up to 1, but that was because, obviously, the spec graph was carefully constructed in a certain way. I am not totally sure how painful would be to normalise properly here, so I will have a think. Worst case scenario we would add this as a "TODO" ticket and pick it up after we merged this PR.